experimental/lib/Transforms/ResourceSharing/Crush.cpp

```diff
@@ -0,0 +1,682 @@
+          CreditBasedSharingPass> {
+
+  CreditBasedSharingPass(StringRef algorithm, StringRef frequencies,
+                         StringRef timingModels, bool firstCFDFC,
+                         double targetCP, unsigned timeout, bool dumpLogs) {
+    this->algorithm = algorithm.str();
+    this->frequencies = frequencies.str();
+    this->timingModels = timingModels.str();
+    this->firstCFDFC = firstCFDFC;
+    this->targetCP = targetCP;
+    this->timeout = timeout;
+    this->dumpLogs = dumpLogs;
+  }
+
+  void runDynamaticPass() override;
+
+  SmallVector<mlir::Operation *> getSharingTargets(handshake::FuncOp funcOp) {
+    SmallVector<Operation *> sharingTargets;
+
+    for (auto &op : funcOp.getOps()) {
+      if (isa<SHARING_TARGETS>(op)) {
+        sharingTargets.emplace_back(&op);
+      }
+    }
+    return sharingTargets;
+  }
+
+  // Call the wrapper class HandshakePlaceBuffersPassWrapper, which again wraps
+  // FPGA20BuffersWrapper
+  LogicalResult runBufferPlacementPass(ModuleOp &modOp, SharingInfo &data) {
+    TimingDatabase timingDB(&getContext());
+    if (failed(TimingDatabase::readFromJSON(timingModels, timingDB)))
+      return failure();
+
+    // running buffer placement on current module
+    mlir::PassManager pm(&getContext());
+    pm.addPass(std::make_unique<HandshakePlaceBuffersPassWrapper>(
+        data, algorithm, frequencies, timingModels, firstCFDFC, targetCP,
+        timeout, dumpLogs));
+    if (failed(pm.run(modOp))) {
+      return failure();
+    }
+
+    return success();
+  }
+};
+} // namespace
+
+// For two sharing groups, check if the following criteria hold (see
+// descriptions below).
+bool checkGroupMergable(const Group &g1, const Group &g2,
+                        FuncPerfInfo funcPerfInfo) {
+  if (g1.empty() || g2.empty())
+    return false;
+
+  std::set<Operation *> gMerged;
+  gMerged.insert(g1.begin(), g1.end());
+  gMerged.insert(g2.begin(), g2.end());
+
+  OperationName opName = (*(gMerged.begin()))->getName();
+
+  // 1. The merged group must have operations of the same type.
+  for (auto *op : gMerged) {
+    if (op->getName() != opName)
+      return false;
+  }
+
+  // 2. For each CFC, the sum of occupancy must be smaller than the capacity
+  // (i.e., units in CFC must no greater than the II).
+  // This is equivalent to checking that throughput * n_ops <= 1;
+
+  // 3. For each CFC, there must be no two operations have the same SCC ID (this
+  // is simplified).
+  for (auto cf : funcPerfInfo.critCfcs) {
+    // for each cf, numOps contains the number of operations
+    // that are in the merged group and also in cf
+    unsigned numOps = 0;
+
+    // listOfSccIds: a list of SCC IDs that the group has
+    // it is used to check if there are any duplicates (i.e.,
+    // two operation that are in the same SCC cannot be in the
+    // same sharing group).
+    std::vector<size_t> listOfSccIds;
+    for (auto *op : (funcPerfInfo.cfUnits)[cf]) {
```

Comment:
nit: no need for parentheses around `funcPerInfo.cfUnits`. Same comment applies below